### PR TITLE
feat(api): add endpoint to list and create events

### PR DESCRIPTION
**Pull request by arminpatel**
Event create access is given to `core members`, `co-coordinators`, and `coordinators`.

### CODE REVIEW

File: api/api/auth/models.py [hunk 1]
```diff
@@ -21,7 +21,7 @@ def find_or_create(user):
     @staticmethod
     def if_expired_get_new(token, user):
         token_life = (datetime.now(timezone.utc) - token.created_at).days
-        if(token_life >= settings.AUTH_ALLOWED_TOKEN_LIFE_IN_DAYS):
+        if (token_life >= settings.AUTH_ALLOWED_TOKEN_LIFE_IN_DAYS):
```
Comment (Samy-33) [inline api/api/auth/models.py:24]: (...) not required.

File: api/api/events/test_list_create_events.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+@pytest.fixture(scope='function')
+def test_user(django_user_model):
+    return django_user_model.objects.create(email='test@user.com',
+                                            first_name='test',
+                                            last_name='User')
+
+
+@pytest.mark.django_db
+@pytest.fixture(scope='function')
+def test_club(django_user_model):
```
Comment (Samy-33) [inline api/api/events/test_list_create_events.py:22]: `django_user_model` is not being used.

File: api/api/events/test_list_create_events.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    response = client.get('/events/', format='json')
+
+    # then
+    assert response.status_code == status.HTTP_200_OK
+    assert response.data['count'] == 1
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('position', ['Core member', 'Coordinator', 'Co-Coordinator'])
+def testCreateEvents_coreMemberRequest_creationSuccessful(client, test_user, test_club, position):
```
Comment (Samy-33) [inline api/api/events/test_list_create_events.py:70]: Should be `testCreateEvent` at all the other places as well.

File: api/api/events/test_list_create_events.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+                             'club': 1,
+                             'registration_approval_required': False,
+                             'address': 'c'}
+    assert created_event.name == 'Introduction to Git and Github'
+
+
+@pytest.mark.django_db
+def testCreateEvents_adminRequest_creationSuccessful(client, test_user, test_club):
+    # given
+    test_user.is_staff = True
```
Comment (Samy-33) [inline api/api/events/test_list_create_events.py:109]: After update, don't we need to save the test_user?

File: api/api/events/test_list_create_events.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+                                        'image_url': 'https://asite.com',
+                                        'address': 'c',
+                                        'club': test_club.id
+                                        })
+    # then
+    assert response.status_code == status.HTTP_403_FORBIDDEN
+
+
+@pytest.mark.django_db
+def testCreateEvents_missingDataInRequest_throwsBasRequest(client, test_user, test_club):
```
Comment (Samy-33) [inline api/api/events/test_list_create_events.py:159]: `throwsBadRequest`

File: api/api/events/test_list_create_events.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+    # request missing start and end time
+    response = client.post('/events/', {'name': 'Introduction to Git and Github',
+                                        'description': 'some desc',
+                                        'image_url': 'https://asite.com',
+                                        'address': 'c',
+                                        'club': test_club.id
+                                        })
+
+    # then
+    assert response.status_code == status.HTTP_400_BAD_REQUEST
```
Comment (Samy-33) [inline api/api/events/test_list_create_events.py:173]: Should we also test, what the error message is? It should clearly state what's missing.

**Review by Samy-33 — CHANGES_REQUESTED** [review]

**Comment by Samy-33** [comment]
Try to break down features into multiple PRs. That way it's easier to review and also one can perform a better review.

File: api/api/events/test_list_create_events.py (hunk 4 above)
Comment (arminpatel) [reply api/api/events/test_list_create_events.py:109]: > After update, don't we need to save the test_user?

It does needs to be saved according to the docs.

File: api/api/events/test_list_create_events.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    # when
+    response = client.get('/events/', format='json')
+
+    # then
+    assert response.status_code == status.HTTP_200_OK
+    assert response.data['count'] == 1
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('position', ['Core member', 'Coordinator', 'Co-Coordinator'])
```
Comment (Samy-33) [inline api/api/events/test_list_create_events.py:69]: Lets use constants defined for these values i.e. `Role.CORE_MEMBER` and so on.

File: api/api/events/views.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+        if self._has_create_permission(serializer.validated_data):
+            serializer.save()
+        else:
+            raise PermissionDenied({"message": "You are not allowed to perform this action"})
+
+    # Allow only Core members (including Coordinator and Co-Coordinator)
+    # and admins to create events
+    def _has_create_permission(self, validated_data):
+        allowed_roles = validated_data['club'].roles.filter(
+                        Q(name='Core member') | Q(name='Coordinator') | Q(name='Co-Coordinator'))
```
Comment (Samy-33) [inline api/api/events/views.py:24]: Same, let's use constants defined for these roles.

**Review by Samy-33 — CHANGES_REQUESTED** [review]

**Review by Samy-33 — APPROVED** [review]
LGTM

**Pull request merged**